src/Orchard.Web/Modules/Orchard.OpenId/Controllers/AccountController.cs

```diff
@@ -81,16 +83,20 @@ public class AccountController : Controller {
             return this.RedirectLocal(returnUrl);
         }
 
-        public void Challenge(string openIdProvider) {
+        [SuppressMessage("Microsoft.Design", "CA1054:UriParametersShouldNotBeStrings", Justification = "Needs to take same parameter type as Controller.Redirect()")]
+        public void Challenge(string openIdProvider, string returnUrl) {
             _userEventHandler.LoggingIn(openIdProvider, String.Empty);
 
             if (String.IsNullOrWhiteSpace(openIdProvider))
                 openIdProvider = OpenIdConnectAuthenticationDefaults.AuthenticationType;
 
             if (Request.IsAuthenticated) {
-                Redirect(Url.Content("~/"));
+                Redirect(returnUrl);
```

Comment:
Don't use `returnUrl` anywhere in a redirect without checking the destination is allowed, like `RedirectLocal` does or if it needs to go to a provider's domain, then check it too.

This comment applies to every change in this PR.